lib/src/main/java/tech/relaycorp/relaydroid/GatewayClientImpl.kt

```diff
@@ -124,24 +143,28 @@ internal constructor(
 
     // Internal
 
-    // TODO: Review bind checks and uniformise gateway exceptions
     internal suspend fun checkForNewMessages() {
-        val wasBound = gwServiceInteractor != null
-        if (!wasBound) {
-            try {
-                bind()
-            } catch (exp: GatewayBindingException) {
-                logger.log(Level.SEVERE, "Could not bind to gateway to receive new messages", exp)
-                return
+        withContext(coroutineContext) {
+            val wasAlreadyBound = gwServiceInteractor != null
+            if (!wasAlreadyBound) {
+                try {
+                    bind()
+                } catch (exp: GatewayBindingException) {
+                    logger.log(Level.SEVERE, "Could not bind to gateway to receive new messages", exp)
+                    return@withContext
+                }
             }
-        }
 
-        receiveMessages
-            .receive()
-            .onEach(incomingMessageChannel::send)
-            .launchIn(CoroutineScope(coroutineContext))
+            try {
+                receiveMessages
+                    .receive()
+                    .collect(incomingMessageChannel::send)
+            } catch (exp: ReceiveMessagesException) {
+                logger.log(Level.SEVERE, "Could not receive new messages", exp)
+            }
 
-        if (!wasBound) unbind()
+            if (!wasAlreadyBound) unbind()
```

Comment:
Is this safe? Isn't it possible that the `wasAlreadyBound` was `false` but something else bound to it in a different thread since we did the check at the start of the function?

Reply:
Good point. This method can be called at a random time, with the app open or closed. There's a chance you tried sending something in between receiving messages. But we definitely should unbind if the app is on the background, and was notified.

I guess developers will always need to have a recover bind routine, with our current API, even without this possible unbind. Not sure what's the optimal approach here, without us starting to handle binding for the developer.